flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala

```diff
@@ -999,6 +1006,121 @@ class Table(
     new OverWindowedTable(this, overWindows.toArray)
   }
 
+  /**
+    * Performs a flatMap operation with an user-defined table function.
+    *
+    * Scala Example:
+    * {{{
+    *   class MyFlatMapFunction extends TableFunction[Row] {
+    *     def eval(str : String) {
+    *       if (str.contains("#")) {
+    *         val splits = str.split("#")
+    *         collect(Row.of(splits(0), splits(1)))
+    *       }
+    *     }
+    *
+    *     def getResultType(signature: Array[Class[_]]): TypeInformation[_] =
+    *       Types.ROW(Types.STRING, Types.STRING)
+    *   }
+    *
+    *   val func = new MyFlatMapFunction
+    *   table.flatMap(func('c)).as('a, 'b)
+    * }}}
+    *
+    * Java Example:
+    * {{{
+    *   class MyFlatMapFunction extends TableFunction<Row> {
+    *     public void eval(String str) {
+    *       if (str.contains("#")) {
+    *         String[] splits = str.split("#");
+    *         collect(Row.of(splits[0], splits[1]));
+    *       }
+    *     }
+    *
+    *     public TypeInformation getResultType(Class[] signature) {
+    *       return Types.ROW(Types.STRING, Types.STRING);
+    *     }
+    *   }
+    *
+    *   TableFunction func = new MyFlatMapFunction();
+    *   tableEnv.registerFunction("func", func);
+    *   table.flatMap("func(c)").as("a, b");
+    * }}}
+    */
+  def flatMap(tableFunction: Expression): Table = {
+    unwrap(tableFunction, tableEnv) match {
+      case _: TableFunctionCall =>
+      case _ => throw new ValidationException("Only TableFunction can be used in flatMap.")
+    }
+
+    // rename output fields names to avoid ambiguous name
+    val originalCall = UserDefinedFunctionUtils.createLogicalFunctionCall(tableEnv, tableFunction)
+    val originalOutputFieldNames = originalCall.output.map(_.name)
+    val usedFieldNames: mutable.HashSet[String] =
+      logicalPlan.output.foldLeft(mutable.HashSet[String]()) { (s, o) => s += o.name }
+
+    val newOutputFieldNames = originalOutputFieldNames.zipWithIndex.map { case (name, index) =>
+      val newName = if (usedFieldNames.contains(name)) {
+        index + "_" + usedFieldNames.mkString("_") + "_" + name
```

Comment:
May be `index + "_" + usedFieldNames.mkString("_") + "_" + name` ->
`index + "_" + usedFieldNames.mkString("_")` is enough， due to `usedFieldNames` contains the `name`, but apeend a name also fine to me, What do you think?

Reply:
OK. Make sense.